torch/csrc/autograd/functions/tensor.h

```diff
@@ -17,6 +18,8 @@ struct Identity : public TraceableFunction {
 
 struct CopyBackwards : public Function {
   virtual variable_list apply(const variable_list& inputs) override;
+
+  at::Type *src_type;
```

Comment:
Type is not enough - in case of CUDA copies you also need to save the source device

Reply:
is this true for the functions implemented (half, float, double, byte, char, short, int, long)?  Those don't change from gpu->cpu or visa versa and on gpu they keep the device, right?  (I agree in general we need to track the device)

Reply:
These casting functions should be ok, but only because we take care to pin engine threads to a single device as long as they are alive (they used to all run on GPU 0 for a long time). Still, `CopyBackwards` sounds like something that should be able to do backwards of `a.copy_(b)` and this can cross device boundaries in any way, and you need to save the device if you want to make it work. It really doesn't cost us anything and I'd rather add it to be safer to save debugging in the future.

Reply:
Sure, I'm actively working on converting this stuff over to ATen, so that should all be fixed up this week anyway.